modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+  if (!is_int($endTime)) {
+    throw new Exception("Invalid Expiration date.");
+  }
+  $now = (int) date(time());
+
+  if ($now > $endTime) {
+    throw new Exception("Subject confirmation is expired.");
+  }
+
+  if (!strcmp($ip, "::1")) {
+    $ip = "127.0.0.1";
+  }
+
+  if (strcmp($address, $ip)) {
+    throw new Exception("Invalid IP address.");
+  }
+
+  return TRUE;
+}
+
+/**
+ * Check if the request is still valid.
+ */
+function _ding_innskraning_verify_date($doc) {
+  try {
+    $conditions = _ding_innskraning_locate_conditions($doc);
+  }
+  catch (Exception $e) {
+    throw new Exception("Exception while locating Conditions");
+  }
+
+  if (!$conditions) {
+    throw new Exception("Unable to locate Conditions");
+  }
+
+  try {
+    $start = $conditions->getAttribute('NotBefore');
+    $end = $conditions->getAttribute('NotOnOrAfter');
+  }
+  catch (Exception $e) {
+    throw new Exception("Exception while locating start or end");
+  }
+
+  if (!$start || !$end) {
+    throw new Exception("Unable to locate start (NotBefore) or end (NotOnOrAfter)");
+  }
+
+  date_default_timezone_set('Atlantic/Reykjavik');
+  $startTime = strtotime($start);
+  $endTime = strtotime($end);
+  if (!is_int($startTime) || !is_int($endTime)) {
+    throw new Exception("Unable to get time from start (NotBefore) or end (NotOnOrAfter)");
+  }
+  $now = (int) date(time());
+
+  $inSpan = $startTime < $now && $now < $endTime;
+
+  if (!$inSpan) {
+    throw new Exception("Response is not within specified timeframe");
+  }
+
+  return TRUE;
+}
+
+/**
+ * Check valid response certificate from Island.is.
+ */
+function _ding_innskraning_verify_cert($objKeyInfo) {
+  $caFile = _ding_innskraning_get_key();
+
+  $caCert = openssl_x509_read($caFile);
+  $caCertParsed = openssl_x509_parse($caCert, TRUE);
+
+  $cert = $objKeyInfo->getX509Certificate();
+  $parsed = openssl_x509_parse($cert);
+
+  date_default_timezone_set('Atlantic/Reykjavik');
+  $dateFrom = (int) date($parsed['validFrom_time_t']);
+  $dateTo = (int) date($parsed['validTo_time_t']);
+  $nowTime = (int) date(time());
+  if ($nowTime < $dateFrom || $nowTime > $dateTo) {
+    throw new Exception("Certificate expired or not valid yet");
+  }
+
+  $serialNumber = $parsed['subject']['serialNumber'];
+  $issuer = $parsed['issuer']['CN'];
+
+  if ($serialNumber != "6503760649") {
+    throw new Exception("Not the correct serial number in the certificate");
+  }
+
+  if ($issuer != "Traustur bunadur") {
+    throw new Exception("Not the correct certificate issuer");
+  }
+  $subjectKey = $caCertParsed['extensions']['subjectKeyIdentifier'];
+  $authKey = $parsed['extensions']['authorityKeyIdentifier'];
+  $authKey = str_replace('keyid:', '', $authKey);
+
+  $authKey = trim($authKey);
+
+  if (strcasecmp($subjectKey, $authKey)) {
+    throw new Exception("Not correct CA");
+  }
+
+  return TRUE;
+}
+
+/**
+ * Get client IP.
+ */
+function _ding_innskraning_get_client_ip() {
```

Comment:
Why use this instead of Drupals own [`ip_address()`](https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/ip_address/7.x)?